### PR TITLE
⬆️ UPGRADE: sqlalchemy v1.4 (v2 API)

### PR DESCRIPTION
**Pull request by chrisjsewell**
~~Currently based on top of #113~~, supersedes #108

Following migration: https://docs.sqlalchemy.org/en/14/changelog/migration_20.html

- Added `SQLALCHEMY_WARN_20` environmental variable
- Use `future=True` for engine and session creation (V1 -> v2 API)
- `query` -> `select` (V1 -> v2 API)
- Rename `models.py` -> `database.py` (models is too generic does not describe the module's purpose)
- Moved `get_session` -> `database.py` (this method can be independent of the container)
- Added mypy extension: https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html

I noted the `count()` method is now a bit more complex, but this is explained in: https://github.com/sqlalchemy/sqlalchemy/issues/6794

A few things to double-check:

- [x] A number of queries used `with_entities`, but I was unclear why this was necessary,
      i.e. I changed `query(Obj)...with_entities(Obj.hashkey)` to `select(Obj.hashkey)..` @giovannipizzi any reason why this would be an issue?

- [x] Vacuum now fails, since it is in a transaction. I've added a workaround, but checked on this in: https://github.com/sqlalchemy/sqlalchemy/discussions/6959#discussioncomment-1251681

### CODE REVIEW

File: disk_objectstore/container.py [hunk 1]
```diff
@@ -1893,10 +1848,15 @@ def _vacuum(self) -> None:
...
         """
-        engine = self._get_cached_session().get_bind()
-        engine.execute("VACUUM")
+        # VACUUM cannot be performed from within a transaction
+        # see: https://github.com/sqlalchemy/sqlalchemy/discussions/6959
+        session = self._get_cached_session()
+        session.execute(text("COMMIT"))
+        session.execute(text("VACUUM"))
+        # ensure sqlalchemy knows to open a new transaction for the next execution
+        session.commit()
```
Comment (ramirezfranciscof) [inline disk_objectstore/container.py:1857]: What is the difference between `session.execute(text("COMMIT"))` and `session.commit()` and why is it necessary to make it differently here?

**Review by ramirezfranciscof — APPROVED** [review]
Everything great here 👍🏽 I just have one question out of curiosity but this is ready to merge for me.

File: disk_objectstore/container.py (hunk 1 above)
Comment (chrisjsewell) [reply disk_objectstore/container.py:1857]: The difference is that this direct execution bypasses sqlalchemy’s auto-transaction mechanics, and stops it from starting a new transaction before executing the vacuum (which cannot be executed within a transaction)

**Comment by chrisjsewell** [comment]
Cheers!

**Pull request merged**